Django Allauth (#201)

gdude2002 opened the pull request:
This pull request implements Django Allauth as described in #201.

Fixes #201.

---

This is a fairly big PR which, I think, may have had a little scope creep. Here's what it changes:

* Add `django-allauth` dependency
* Implements Discord OAuth logins using `django-allauth`
* Exposes messages from the Django messages framework as slide-in Bulma notifications at the top right of the page, positioned just below the navbar
* Re-styles Bulma notifications when used in the context of the Django messages framework
* Redefines Bulma's bundled colours, in order to make them less loud and make Bulma calculate colours based on the primary blurple better (by redefining turquoise to blurple)
* Implements some logic to synchronise Discord roles and Django permissions groups based on signals emitted by various events

---

I would super love some reviews for this, particularly when it comes to the tests I wrote for the signal handlers, and the signal handling itself.

I'm not super happy with the signal handlers test - it's more setup than test! If anyone has any ideas on how to clean that up, I'd appreciate it.

---

Old TODO:

- [x] Set up signal listeners for Allauth events
- [x] Set up signal listeners for relevant Model save events
- [x] Document the signal system in comments/docstrings
- [x] Fix up message system
    - Move messages to a corner
    - Restyle messages
        - Discord embed-style messages with the left border?
        - More muted colours? Perhaps shades of the primary blurple.
- [ ] ~~Profile page with GH connect/disconnect and account delete~~
- [ ] ~~Create templates for the Allauth-provided account pages~~
- [x] Review the Allauth URLs to ensure we're only using what we need
    - ~~Ensure login page explains the data we're collecting and collects explicit consent~~
- [x] Create a user profile page (or just send them back to the homepage)
- [x] Figure out how to signpost/link to this system from the rest of the site

gdude2002 committed:
Initial Allauth dependency and settings/urls

gdude2002 committed:
Remove accidental static files commit; fix Allauth

gdude2002 committed:
Re-add erroneously removed static files

gdude2002 committed:
Disable Allauth email verification

gdude2002 committed:
Add model to map Discord roles to Django permissions groups

gdude2002 committed:
Remove app_label from role mapping model

Turns out, setting this *does* move the model to a different category under the admin - but it also reparents it to the referenced app, so the migrations are created under that app instead. Nasty.

gdude2002 committed:
Hook up Allauth and model signals to handle group assignments

gdude2002 committed:
Merge branch 'master' into #201-django-allauth

gdude2002 committed:
Tests for signal handlers

gdude2002 committed:
Add docstrings to tests, and test role mapping __str__()

gdude2002 committed:
Clean up and comment signal tests

gdude2002 committed:
Add MessageRedirectView and show messages on the index page

gdude2002 committed:
Set up needed Allauth URLs and add login/logout to navbar

gdude2002 committed:
Test allauth redirect URLs

gdude2002 committed:
Move messages to the top right

gdude2002 committed:
Message styling (default to light for now)

gdude2002 committed:
Add missing CSS for notifications

gdude2002 committed:
Make the Bulma colours less saturated

gdude2002 committed:
Login page and styling fixes

gdude2002 committed:
Replace card on login page with notification

gdude2002 committed:
Prevent saving emails, remove login page

gdude2002 committed:
Remove GH login implementation, out of scope of this PR

gdude2002 committed:
Merge branch 'master' into #201-django-allauth

# Conflicts:
#	Pipfile
#	Pipfile.lock

gdude2002 committed:
Re-lock Pipfile due to merge with master

gdude2002 committed:
Register GH provider manually in tests

jchristgit commented on pydis_site/apps/home/apps.py:
```diff
@@ -1,7 +1,38 @@
+            data["email"] = ""
+            return data
+
+        Provider.extract_extra_data = extract_extra_data
```
shouldn't this subclass, call super(), and then strip the email, in case allauth does something useful here?

gdude2002 replied:
Allauth actually does nothing at all here, just `return data`. I figured this would be better for future work, since you can't actually tell Allauth which class to use - although I guess I could replace the entire class? I feel like that'd break things, though. I dunno.

What do you think?

jchristgit replied:
Yeah, that makes sense.

jchristgit commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,159 @@
+    supposedly because Signals can change in the future and your receivers should still work.
+
+    Signals do provide a list of their arguments when they're initially constructed, but this
+    is purely for documentation purposes only and Django does not enforce it.
```
It's nice that you document this, but I think it would be better to just provide a link to the Django documentation for these three paragraphs

gdude2002 replied:
I agree with you, @lemonsaurus asked me to document it though.

lemonsaurus replied:
There is a link to the Django docs, too. This is fine.

jchristgit commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,159 @@
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(user, account, True)
```
The kwarg is a bit hard to figure out without reading the function, would be nice to

```suggestion
        self._apply_groups(user, account, deletion=True)
```

gdude2002 committed:
Clarify _apply_groups kwarg (Thanks, @jchristgit)

Co-Authored-By: Johannes Christ <jc@jchri.st>

jchristgit commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,159 @@
+                mapping.group for mapping in mappings if mapping.group not in new_groups
+            ]
+
+            add_groups = [group for group in new_groups if group not in current_groups]
```
Looking at the `.remove` call above, can't we have these as generators?

jchristgit replied:
Django evaluates this as a list by itself:

```py
        def add(self, *objs, bulk=True):
            self._remove_prefetched_objects()
            objs = list(objs)
            # ...
```
via `django/db/models/fields/related_descriptors.py`

gdude2002 replied:
This has proven to be not the case.

![image](https://user-images.githubusercontent.com/204153/66676577-c0128c00-ec5f-11e9-8f14-abe1f2241783.png)

jchristgit commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,159 @@
+                account.user.groups.remove(*remove_groups)
+
+            if add_groups:
+                account.user.groups.add(*add_groups)
```
Not fully convinced you need the `if` conditions here. I don't think it hurts to always run this and keep the code simpler

jchristgit commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,319 @@
+        """Test application of groups by role, relating to an admin user."""
+        handler = SignalListener()
+
+        self.assertTrue(self.django_user_discordless.groups.all().count() == 0)
```
this should use `assertEqual(..., 0)`

jchristgit commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,319 @@
+
+        # Remove groups based on the user apparently leaving the server
+        handler._apply_groups(self.discord_admin, self.social_admin, True)
+        self.assertTrue(self.django_user_discordless.groups.all().count() == 0)
```
this should use `assertEqual(..., 0)`

jchristgit commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,319 @@
+
+        # No groups should be applied when the user isn't in the guild
+        handler._apply_groups(self.discord_not_in_guild, self.social_user)
+        self.assertTrue(self.django_user.groups.all().count() == 0)
```
I'm not gonna comment on each, but you know the drill

jchristgit commented on pydis_site/templates/base/navbar.html:
```diff
@@ -83,6 +84,25 @@
+                     href="{% provider_login_url provider.id process="login" scope=scope auth_params=auth_params %}"
+                  >Login with {{ provider.name }}</a>
+              {% endif %}
+            {% endfor %}
```
why not just display the single login url instead of iterating over each until you find discord?

gdude2002 replied:
As far as I can tell, there are a couple reasons to do this:

* There might be more than one Discord OAuth app to login with
* Discord OAuth might not be configured yet

Those aren't necessarily real concerns for the time being, I guess. I do still need to get the URL from the provider though, and this is the official way Allauth provides to do it.

jchristgit reviewed (changes requested):
This looks pretty solid so I will have to resort to notpicking :(

The role mapping model is just about the cleanest way I have seen to accomplish this. I think in my last implementation of this I just hardcoded that, but a nice admin frontend for it is pretty brilliant.

That said, if the role mappings update, shouldn't the signal handler fire?

gdude2002 committed:
Simplify signals.py as per @jchristgit's review

gdude2002 committed:
Clean up signal tests as per @jchristgit's review

gdude2002 committed:
Signals: `add()` does require a list, a generator won't do.

gdude2002 committed:
Signals: Fix test for `deletion=True` change

gdude2002 committed:
Signals: Handle (and test) mapping updates/deletions

This also enforces unique values for both attributes on the RoleMapping model. Supporting configurations where this isn't the case would introduce quite a lot of added complexity.

gdude2002 committed:
Merge branch 'master' into #201-django-allauth

gdude2002 committed:
Merge branch 'master' into #201-django-allauth

jchristgit reviewed (approved):
very solid work

lemonsaurus commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,192 @@
+from pydis_site.apps.staff.models import RoleMapping
+
+
+class SignalListener:
```
This class name is too generic. We might have other signal listener classes other places in the site, so it would probably be good if we make this class name more specific. something like `AllauthSignalListener` would be fine.

Besides, SignalListener sounds a bit like a Django internal, and it feels like it should take arguments that configure which signals to listen to, or something.

lemonsaurus commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,192 @@
+        user_logged_in.connect(self.user_logged_in)
+
+    def user_logged_in(self, sender: Type[DjangoUser], **kwargs) -> None:
+        """Handles signals relating to Allauth logins."""
```
I think you can probably improve these docstrings -- that goes for pretty much all the docstrings in this file. I understand that it handles these signals, but what does it actually _do_ when the user logs in?

lemonsaurus commented on pydis_site/apps/home/signals.py:
```diff
@@ -0,0 +1,192 @@
+
+        self._apply_groups(instance, account)
+
+    def _apply_groups(
```
Don't we lint for docstrings on functions like this?

Either way I think there should be a docstring.

gdude2002 replied:
We don't lint for docstrings on private methods - just public ones - but I'm happy to add one.

lemonsaurus commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,392 @@
+class SignalListenerTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        """Executed when testing begins."""
```
Again, this docstring could stand to be improved, especially considering the complexity of this setup method. 

What are we setting up? Why?

lemonsaurus commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,392 @@
+                instance=self.discord_user,
+                raw=True,
+                created=None,  # Not realistic, but we don't use it
+                using=None,  # Again, we don't use it
```
can we line these inlines up? It'll probably look better, and PEP8 allows it.

gdude2002 replied:
Do you mean, like, so they're not all on separate lines?

lemonsaurus replied:
no no, just the inline comments.

```py
                created=None,  # Not realistic, but we don't use it
                using=None,    # Again, we don't use it
```

lemonsaurus commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,392 @@
+                instance=self.discord_user,
+                raw=False,
+                created=None,  # Not realistic, but we don't use it
+                using=None,  # Again, we don't use it
```
same thing here

lemonsaurus commented on pydis_site/apps/home/tests/test_signal_listener.py:
```diff
@@ -0,0 +1,392 @@
+        mock_obj = mock.Mock()
+
+        with mock.patch.object(SignalListener, "_apply_groups", mock_obj):
+            _ = SignalListener()
```
Would it work to just not store this at all? like just doing `SignalListener()`? I'm mostly curious, the way you've done it here doesn't really bother me, but it might be cleaner to just call it like a function if it actually works.

gdude2002 replied:
I kind of just pre-emptively did this to make sure garbage collection didn't screw us over, but I can check.

gdude2002 replied:
Works fine without that, yep.

lemonsaurus commented on pydis_site/apps/home/tests/test_views.py:
```diff
@@ -7,3 +7,17 @@ def test_index_returns_200(self):
+
+
+class TestLoginCancelledReturns302(TestCase):
+    def test_login_cancelled_returns_302(self):
```
can we docstring these? It's nice because when the tests are run, it will display the docstrings instead of the method names as it lists them, and that looks nice. I know the function name is self-documenting, but I'd rather read a natural language sentence that describes the same thing.

No need for a long docstring here, just a one-liner is fine.

same with the method below.

lemonsaurus commented on pydis_site/apps/home/urls.py:
```diff
@@ -1,14 +1,36 @@
+    ),
+    path(
+        'accounts/login/error', MessageRedirectView.as_view(
+            pattern_name="home", message="Login failed due to an error, please try again.",
```
"Failed due to an error" is a bit of a pleonasm. Maybe "The Login process encountered an error" or something?

lemonsaurus commented on pydis_site/settings.py:
```diff
@@ -278,9 +294,18 @@
+        "primary": "#7289DA",  # Discord blurple
+
+        "orange": "#ffb39b",  # Bulma default, but at a saturation of 100
+        "yellow": "#ffea9b",  # Bulma default, but at a saturation of 100
```
can we line these up?

gdude2002 replied:
Not sure what you mean by lining up here either

lemonsaurus replied:
```py
thing()  # this
with.fire()  # looks
pull_chain()  # messy

thing()       # this
with.fire()   # looks
pull_chain()  # clean
```

lemonsaurus reviewed (dismissed):
Very high quality PR, but I do have a few requests before I can approve.

gdude2002 committed:
Allauth: Address reviews by @lemonsaurus

gdude2002 committed:
Merge branch 'master' into #201-django-allauth

scragly reviewed (approved):
Absolutely sexy. LGTM
